### ISSUE TITLE
Config-schema.json does not contain type property for optional configurables

### ISSUE THREAD

**Issue by udyogicx** [ballerina-platform/ballerina-lang#34417]
**Description:**
If the configurable variable is optional, config-schema does not created with the type property.

config-schema.json obtained for `gsheet_row_update_to_gmail` sample:
```
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "type": "object",
  "properties": {
    "udyogimunasinghe": {
      "type": "object",
      "properties": {
        "google_sheets_row_update_to_gmail_message": {
          "type": "object",
          "properties": {
            "sheetOAuthConfig": {
              "type": "object",
              "properties": {
                "refreshUrl": {
                  "type": "string"
                },
                "refreshToken": {
                  "type": "string"
                },
                "clientId": {
                  "type": "string"
                },
                "clientSecret": {
                  "type": "string"
                },
                "scopes": {},
                "defaultTokenExpTime": {
                  "type": "number"
                },
                "clockSkew": {
                  "type": "number"
                },
                "optionalParams": {},
                "credentialBearer": {},
                "clientConfig": {}
              },
              "additionalProperties": false,
              "description": ""
            },
            "spreadsheetId": {
              "type": "string",
              "description": ""
            },
            "worksheetName": {
              "type": "string",
              "description": ""
            },
            "gmailOauthConfig": {
              "type": "object",
              "properties": {
                "refreshUrl": {
                  "type": "string"
                },
                "refreshToken": {
                  "type": "string"
                },
                "clientId": {
                  "type": "string"
                },
                "clientSecret": {
                  "type": "string"
                },
                "scopes": {},
                "defaultTokenExpTime": {
                  "type": "number"
                },
                "clockSkew": {
                  "type": "number"
                },
                "optionalParams": {},
                "credentialBearer": {},
                "clientConfig": {}
              },
              "additionalProperties": false,
              "description": ""
            },
            "recipientAddress": {
              "type": "string",
              "description": ""
            },
            "senderAddress": {
              "type": "string",
              "description": ""
            }
          },
          "additionalProperties": false,
          "required": [
            "sheetOAuthConfig",
            "spreadsheetId",
            "worksheetName",
            "gmailOauthConfig",
            "recipientAddress",
            "senderAddress"
          ]
        }
      },
      "additionalProperties": false
    }
  },
  "additionalProperties": false
}
```

According to [OAuth2RefreshTokenGrantConfig](https://lib.ballerina.io/ballerina/http/2.1.0/records/OAuth2RefreshTokenGrantConfig) and [RefreshTokenGrantConfig](https://lib.ballerina.io/ballerina/oauth2/2.1.0/records/RefreshTokenGrantConfig), `scopes`, `optionalParams`, `credentialBearer` and `clientConfig` are optional fields and the type for those fields are not set in config-schema.json.

**Steps to reproduce:**
Use [gsheet_row_update_to_gmail](https://central.ballerina.io/choreo/gsheet_row_update_to_gmail) sample and build it with `bal build --generate-config-schema`

**Affected Versions:**
slbeta6

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

**Comment by udyogicx** [comment]
Additonally, inside record type configurables, there should be a field called `required` to include the required fields of the record.

Ex: In the config-schema.json provided above, under `sheetOAuthConfig`, there should be a field as `required` and it should contain all the required fields of the record.

**Comment by sachinira** [comment]
All the complex configurable issues observed in the sample codes are listed [here](https://docs.google.com/spreadsheets/d/1Xa8vluggZZsr9a6mvs2HIZNliW3FNHcV-8um_HL2gp4/edit?usp=sharing)

**Comment by Dilhasha** [comment]
Following are the identified limitations.

- Improvement to union type to support the union of complex types(e.g. records, arrays)
- Add required field support for record types

When adding the above fix, noticed the following issue.

[http:OAuth2RefreshTokenGrantConfig](https://lib.ballerina.io/ballerina/http/2.0.0/records/OAuth2RefreshTokenGrantConfig) contains fields that recursively refer the same record i.e. [oauth2:ClientConfiguration](https://lib.ballerina.io/ballerina/oauth2/2.0.0/records/ClientConfiguration). Due to this the structures fail to complete due to a deadlock situation. Even though this is allowed in Ballerina, there is no way to represent this in JSON schema.

In order to resolve the above, we have maintained already visited structures for complex types and if there is a reference to an incomplete structure, we set the type as `null`. So, if you notice a `null` type, it is due to a recursive reference.
